src/czar/HttpCzarIngestModuleBase.cc

```diff
@@ -0,0 +1,374 @@
+// In reality the parameters are not needed for ingesting regular tables which is
+// is the only kind of tables supported by the current implementation of the module.
+// Some value of the parameters are still required by the Replication/Ingest system's API.
+unsigned int defaultNumStripes = 340;
+unsigned int defaultNumSubStripes = 3;
+float defaultOverlap = 0.01667;
+
+string const defaultDirectorTableName = "qserv_director";
+unsigned int const defaultChunkId = 0;
+
+}  // namespace
+
+namespace lsst::qserv::czar {
+
+HttpCzarIngestModuleBase::HttpCzarIngestModuleBase(asio::io_service& io_service)
+        : _io_service(io_service),
+          _registryBaseUrl("http://" + cconfig::CzarConfig::instance()->replicationRegistryHost() + ":" +
+                           to_string(cconfig::CzarConfig::instance()->replicationRegistryPort())) {}
+
+list<pair<string, string>> HttpCzarIngestModuleBase::ingestData(
+        string const& databaseName, string const& tableName, json const& schema, json const& indexes,
+        function<map<string, string>(uint32_t)> const& submitRequestsToWorkers) {
+    _unpublishOrCreateDatabase(databaseName);
+    _createTable(databaseName, tableName, schema);
+
+    uint32_t transactionId = 0;
+    try {
+        transactionId = _startTransaction(databaseName);
+    } catch (exception const& ex) {
+        deleteTable(databaseName, tableName);
+        throw;
+    }
+
+    map<string, string> const workerErrors = submitRequestsToWorkers(transactionId);
+
+    if (!workerErrors.empty()) {
+        _abortTransaction(transactionId);
+        deleteTable(databaseName, tableName);
+        json const errorExt = json::object({{"worker_errors", workerErrors}});
+        throw http::Error(__func__, "error(s) reported by workers", errorExt);
+    }
+
+    // Success: commit the transaction and publish the database.
+    _commitTransaction(transactionId);
+    _publishDatabase(databaseName);
+
+    // The post-ingest steps are optional. They are allowed to fail without affecting the success
+    // of the ingest. A warning will be reported in the response in case of a failure.
+    list<pair<string, string>> warnings;
+    _createIndexes(__func__, databaseName, tableName, indexes, warnings);
+    _countRows(__func__, databaseName, tableName, warnings);
+    return warnings;
+}
+
+void HttpCzarIngestModuleBase::verifyUserDatabaseName(string const& func, string const& databaseName) const {
+    string const userDatabaseNamesPrefix = "user_";
+    if ((databaseName.size() <= userDatabaseNamesPrefix.size()) ||
+        !boost::iequals(databaseName.substr(0, userDatabaseNamesPrefix.size()), userDatabaseNamesPrefix)) {
+        auto err = "database name doesn't start with the prefix: " + userDatabaseNamesPrefix;
+        throw http::Error(func, err);
+    }
+}
+
+void HttpCzarIngestModuleBase::verifyUserTableName(string const& func, string const& tableName) const {
+    string const qservTableNamesPrefix = "qserv";
```

Comment:
The documentation said "qserv_" was reserved, although having "qserv" be reserved probably makes more sense.

Reply:
Probably